nrf24L01_code/bootloader/WriteApp.c

```diff
@@ -0,0 +1,383 @@
+#define BOOTLOADER_FLASH_ERROR		((uint8_t) 'L')
+
+
+// Flashing commands
+#define COMM_GET_STATE				((uint8_t) 'q')
+#define COMM_START_FLASH_PROCEDURE	((uint8_t) 'w')
+#define COMM_SEND_CHUNCK			((uint8_t) 'e')
+#define COMM_FLASH_PAGE				((uint8_t) 'r')
+#define COMM_NEXT_PAGE				((uint8_t) 't')
+#define COMM_CHECK_CKS				((uint8_t) 'y')
+#define COMM_FINISH_FLASH			((uint8_t) 'u')
+
+
+
+#define COMM_ABORT					((uint8_t) 'o')
+#define COMM_NO_COMMAND				((uint8_t) 'p')
+
+char bootloader_state = BOOTLOADER_INIT_NRF;
+
+
+typedef struct rxDataStruct
+{
+	uint8_t data[BUFFER_LENGTH];
+	uint8_t command;
+	uint8_t length;
+	uint8_t hasMessage;
+}rxDataStruct;
+
+rxDataStruct rxData;
+uint8_t flashPipe = 0;
+
+
+void rx_handler(uint8_t pipe, uint8_t * data, uint8_t payload_length) {
+	rxData.command = data[0];
+	memcpy(rxData.data, data + 1, payload_length - 1);	
+	rxData.length = payload_length - 1;	
+	rxData.hasMessage = MSG_RECEIVED;
+}
+void tx_handler(radio_tx_status tx_status) {
+	
+}
+
+
+
+void initNrf(uint8_t * rx_address)
+{
+	radio_config cfg = {
+		RADIO_ADDRESS_5BYTES,
+		RADIO_RETRANSMIT_WAIT_3000US,
+		RADIO_RETRANSMIT_15,
+		CHANNEL_112,
+		RADIO_250KBIT,
+		RADIO_CRC2_ENABLED,
+		RADIO_COUNT_WAVE_DISABLED,
+		RADIO_HIGHEST_0DBM,
+		RADIO_DYNAMIC_PAYLOAD_ENABLED,
+		RADIO_ACK_PAYLOAD_ENABLED,
+		RADIO_DYNAMIC_ACK_DISABLED,
+		RADIO_BOOTLOADER
+	};
+	__nrfRadio_Init(cfg);
+	
+	pipe_config pipe_cfg0 =
+	{
+		RADIO_PIPE0,
+		rx_address,
+		5,
+		RADIO_PIPE_RX_ENABLED,
+		RADIO_PIPE_AA_ENABLED,
+		RADIO_PIPE_DYNAMIC_PYALOAD_ENABLED
+	};
+	__nrfRadio_PipeConfig(pipe_cfg0);
+	
+	__nrfRadio_SetRxCallback(rx_handler);
+	__nrfRadio_SetTxCallback(tx_handler);
+	__nrfRadio_PowerUp();
+	__nrfRadio_ListeningMode();
+}
+void waitRx()
+{
+	uint32_t timeoutCounter;
+	timeoutCounter = MSG_REC_TIMEOUT;
+	rxData.hasMessage = MSG_NO_DATA;	
+	rxData.command = COMM_NO_COMMAND;
+	do{
+		timeoutCounter--;
+		__nrfRadio_Main();
+		_delay_us(10);
+	}while (!rxData.hasMessage && (timeoutCounter > 0));
+	
+	if (!rxData.hasMessage || (timeoutCounter == 0))
+		rxData.hasMessage = MSG_NO_DATA;		
+}
+
+
+// Bootloader flash
+void boot_program_page (uint32_t page, uint8_t *buf)
+{
+	uint16_t i;
+	uint8_t sreg;
+	// Disable interrupts.
+	sreg = SREG;
+	cli();
+	eeprom_busy_wait ();
+	boot_page_erase (page);
+	boot_spm_busy_wait (); // Wait until the memory is erased.
+	for (i=0; i<SPM_PAGESIZE; i+=2)
+	{
+		// Set up little-endian word.
+		uint16_t w = *buf++;
+		w += (*buf++) << 8;
+		boot_page_fill (page + i, w);
+	}
+	boot_page_write (page); // Store buffer in flash page.
+	boot_spm_busy_wait(); // Wait until the memory is written.
+	// Reenable RWW-section again. We need this if we want to jump back
+	// to the application after bootloading.
+	boot_rww_enable ();
+	// Re-enable interrupts (if they were ever enabled).
+	SREG = sreg;
+}
+
+
+void startFlash(uint8_t * rx_address)
+{
+	uint8_t ackResponse [BUFFER_LENGTH];			
```

Comment:
Same here, I think we should initialize this (ackResponse) to 0

Reply:
Also BUFFER_LENGTH is 32 bytes, is not a good idea to send 32 Bytes from which just one byte is used e.g we answer with 'q' and 31 empty bytes. We should use at least 1 or 2 bytes, this will speed up the transmission because the NRF is supporting the dynamic payload and the message can be 1B up to 32B, so make no sense to send 32B in this case.